testng-core/src/test/java/test/listeners/ListenersTest.java

```diff
@@ -92,6 +93,35 @@ public void testWiringInOfListenersInMultipleTestTagsWithListenerInSuite() {
     assertThat(logs.get("Xml_Test_2")).containsAll(expected);
   }
 
+  @Test(description = "GITHUB-2796")
+  public void testCorrectOrderOfOnBeforeClassAndOnAfterClass() {
+    setupTest(true);
+    Map<String, Long> logs = ListenerSample.getTimeLogs();
```

Comment:
Since this would give us a `LinkedHashMap` wherein the insertion order would be maintained, we could just retrieve the keys from the map and then run an assertion that compares expectedOrder with the retrieved keys and ensure that they both are in the same order. The test would be a bit more succinct and concise with that change and wouldnt require the iteration or usage of `SoftAssertions`

Reply:
I like the idea of this but, unless I'm missing something, that'd only allow the verification of the `ListenerSample` order, and not the `ListenerSample` and `TestClassSample` combined order. Note the following line where I add all the time logs from `TestClassSample` to this same map before checking the order. I'm unaware of a data structure that will be able to combine while maintaining the order based on the time originally added. I would definitely like to use a better solution that doesn't involve a bunch of 1ms waits combined with currentTimeMillis, but I wasn't able to think of one.

Reply:
@RSXEnthusiast - Ok, what if you did something like this

This would be a new test class and you shouldn't be using the same test class

```java
package test.listeners.issue2796;

import java.time.Instant;
import java.util.LinkedHashSet;
import java.util.Objects;
import java.util.Set;
import org.testng.IClassListener;
import org.testng.ITestClass;
import org.testng.annotations.AfterClass;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.Listeners;
import org.testng.annotations.Test;
import test.listeners.issue2796.TestClassSample.LocalLogs;

@Listeners(LocalLogs.class)
public class TestClassSample {

  @BeforeClass
  public void beforeClass() throws InterruptedException {
    TestLogAppender.addLog("beforeClass");
  }

  @Test
  public void testMethod() throws InterruptedException {
  }

  @AfterClass
  public void afterClass() throws InterruptedException {
    TestLogAppender.addLog("afterClass");
  }
  public static class LocalLogs implements IClassListener {

    @Override
    public void onBeforeClass(ITestClass testClass) {
      TestLogAppender.addLog("onBeforeClass");
    }

    @Override
    public void onAfterClass(ITestClass testClass) {
      TestLogAppender.addLog("onAfterClass");
    }
  }

  public static class TestLogAppender {

    private static final Set<DataItem> logs = new LinkedHashSet<>();

    public static Set<DataItem> getLogs() {
      return logs;
    }

    public static synchronized void addLog(String method) {
      logs.add(new DataItem(method));
    }
  }

  public static class DataItem {
    private final String methodName;
    private final long timestamp;

    public DataItem(String methodName) {
      this.methodName = methodName;
      this.timestamp = Instant.now().getNano();
    }

    public String getMethodName() {
      return methodName;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (o == null || getClass() != o.getClass()) {
        return false;
      }
      DataItem dataItem = (DataItem) o;
      return timestamp == dataItem.timestamp;
    }

    @Override
    public int hashCode() {
      return Objects.hash(timestamp);
    }
  }
}

```

and this would be your new test method

```java
  @Test(description = "GITHUB-2796")
  public void testCorrectOrderOfOnBeforeClassAndOnAfterClass() {
    TestNG testng = new TestNG();
    XmlSuite xmlSuite = createXmlSuite("Xml_Suite");
    createXmlTest(xmlSuite, "Xml_Test_1", test.listeners.issue2796.TestClassSample.class);
    testng.setXmlSuites(Collections.singletonList(xmlSuite));
    testng.setVerbose(2);
    testng.run();
    String[] logs = TestLogAppender.getLogs().stream()
        .map(DataItem::getMethodName)
        .toArray(String[]::new);
    String[] expectedOrder = new String[]{
            "onBeforeClass",
            "beforeClass",
            "afterClass",
            "onAfterClass"
    };
    assertThat(logs).containsExactly(expectedOrder);
  }
```

Reply:
I was able to simplify it from this, but great ideas!